fabtests/prov/efa/src/rdm_rnr_queue_resend.c

```diff
@@ -0,0 +1,438 @@
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * This test resets RNR retry counter to 0 via fi_setopt, and check if
+ * queue/re-send logic works correctly for different packet types. To run
+ * the test, one needs to use `-c` option to specify the category of packet types
+ *
+ *    The tests against all different packet types can be running with runfabtests.sh.
+ *    The relationship between each test and the corresponding packet type it checks against is:
+ *    "fi_efa_rnr_queue_resend -c 0 -S 1048576" - CTS, DATA / EOR(RDMA Read)
+ *    "fi_efa_rnr_queue_resend -c 0 -o read -S 4" - READRSP
+ *    "fi_efa_rnr_queue_resend -c 0 -A read -S 4" - ATOMRSP
+ *    "fi_efa_rnr_queue_resend -c 0 -U -S 4" - RECEIPT
+ *    "fi_efa_rnr_queue_resend -c 1 -S 4" - EAGER_MSGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -T -S 4" - EAGER_TAGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -S 16384" - MEDIUM_MSGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -T -S 16384" - MEDIUM_TAGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -S 1048576" - LONGCTS_MSGRTM / LONGREAD_MSGRTM (RDMA Read)
+ *    "fi_efa_rnr_queue_resend -c 1 -T -S 1048576" - LONGCTS_TAGRTM / LONGREAD_TAGRTM (RDMA Read)
+ *    "fi_efa_rnr_queue_resend -c 1 -o write -S 4" - EAGER_RTW
+ *    "fi_efa_rnr_queue_resend -c 1 -o write -S 1048576" - LONGCTS_RTW / LONGREAD_RTW (RDMA Read)
+ *    "fi_efa_rnr_queue_resend -c 1 -o read -S 4" - SHORT_RTR
+ *    "fi_efa_rnr_queue_resend -c 1 -o read -S 1048576" - LONGCTS_RTR
+ *    "fi_efa_rnr_queue_resend -c 1 -A write -S 4" - WRITE_RTA
+ *    "fi_efa_rnr_queue_resend -c 1 -A read -S 4" - FETCH_RTA
+ *    "fi_efa_rnr_queue_resend -c 1 -A cswap -S 4" - COMPARE_RTA
+ *    "fi_efa_rnr_queue_resend -c 1 -U -S 4" - DC_EAGER_MSGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -T -U -S 4" - DC_EAGER_TAGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -U -S 16384" - DC_MEDIUM_MSGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -T -U -S 16384" - DC_MEDIUM_TAGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -U -S 1048576" - DC_LONGCTS_MSGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -T -U -S 1048576" - DC_LONGCTS_TAGRTM
+ *    "fi_efa_rnr_queue_resend -c 1 -o write -U -S 4" - DC_EAGER_RTW
+ *    "fi_efa_rnr_queue_resend -c 1 -o write -U -S 1048576" - DC_LONGCTS_RTW
+ *    "fi_efa_rnr_queue_resend -c 1 -A write -U -S 4" - DC_WRITE_RTA
+ *
+ * In addition, HANDSHAKE packet's queue/re-send can be easily triggered during
+ * initial ft_sync's ft_rx() on the server side, as the client does not
+ * pre-post internal rx buffers until it polls completion in ft_sync's
+ * ft_tx(). All of the above tests have the sync procedure, so we do not
+ * add a dedicated test to check HANDSHAKE packet.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <rdma/fi_atomic.h>
+#include <shared.h>
+#include "efa_rnr_shared.h"
+
+int expected_rnr_error=1;
```

Comment:
Since this is a global variable, can you make the name special like `global_expected_rnr_error`?

Reply:
Done.